Recipe describing how to add external identifiers to Wikidata (#424)
FAIRplus/the-fair-cookbook: content/recipes/interoperability/registering-swisslipids-identifiers.md

# Registering SwissLipids identifiers in Wikidata

 ````{panels_fairplus}
:identifier_text: FCBxxx
:identifier_link: 'https://w3id.org/faircookbook/FCBxxx'
:difficulty_level: 5
:recipe_type: hands_on
:reading_time_minutes: 15

> :intended_audience: bioinformatician, data_scientist, data_engineer
> :has_executable_code: nope

Reviewer: ```suggestion
:intended_audience: bioinformatician, data_scientist, data_engineer
:maturity_level: 1
:maturity_indicator: 8
:has_executable_code: nope
```

:recipe_name: Registering SwissLipids identifiers in Wikidata
```` 

## Main Objectives

> The objective of this recipe is to describe how external identifiers can be added to Wikidata in an automated.

Reviewer: ```suggestion
The objective of this recipe is to describe how external identifiers can be added to Wikidata in an automated way.
```

___

## Requirements

To record identifiers of some resource to make the content more findable and more interoperable, first a [Wikidata](https://www.wikidata.org/) property
needs to be requested. That process is described [here](https://www.wikidata.org/wiki/Wikidata:Property_proposal).

---

## Main Content

> Making datasets more findable and databases more interoperable, it is recommended to share not just top-level metadata, but also record-level

Reviewer: ```suggestion
To make datasets more findable and databases more interoperable, it is recommended to share not just top-level metadata, but also record-level
```
```suggestion
To make datasets more findable and databases more interoperable, it is recommended to share not just top-level metadata, but also record-level
```

> metadata. For example, for chemical databases, you can share your chemical compound identifiers as open data, while keeping the data itself
> in your own database. That allows other databases, like Wikidata, to link to your database. First, that makes your data more findable,
> but it also makes your data easier to integrate and therefore more interoperable.

Reviewer: ```suggestion
metadata. 
For example, for chemical databases, one may share the chemical compound identifiers as open data, while keeping the data itself in their own database. 
That allows other resources, like Wikidata, to link to a private database.
First, that makes such data more findable, but it also makes the same data easier to integrate, and therefore more interoperable.
```

> This recipe follows this idea and starts with a CCZero-licensed mapping file, linking InChIKeys and database identifiers.

Reviewer: ```suggestion
This recipe follows this idea and starts with a [CC0](https://creativecommons.org/publicdomain/zero/1.0/)-licensed mapping file, linking InChIKeys and database identifiers.
```

Recipe *InChI and SMILES identifiers for chemical structures* ([fcb:FCB007](https://w3id.org/faircookbook/FCB007)) explains how InChIKeys
can be generated for chemical compounds in your database.

As an example, this recipe demonstrates the approach of listing [SwissLipid](http://www.swisslipids.org/#/) {footcite}`Aimo2015SwissLipids`

> identifiers in Wikidata, developed at the 2021 BioHackathon Europe. Earlier, a [Swiss Lipid property for Wikidata](https://www.wikidata.org/wiki/Wikidata:Property_proposal/SwissLipids_identifier)

Reviewer: ```suggestion
identifiers in Wikidata, developed at the 2021 BioHackathon Europe. 
Earlier, a [Swiss Lipid property for Wikidata](https://www.wikidata.org/wiki/Wikidata:Property_proposal/SwissLipids_identifier)
```

was already proposed, later approved, and created in Wikidata [just before the 2020 BioHackathon Europe](https://www.wikidata.org/w/index.php?title=Property:P8691&oldid=1287579005).
An existing Wikidata property is a requirement to adding the external identifiers.

The recipe uses [Bacting](https://github.com/egonw/bacting) {footcite}`Willighagen2021` which can be used in Groovy and Python. We
will use the first here. The starting point is mappings (tuples) that link the SwissLipid identifiers to InChIKeys. The latter
is used to find the matching Wikidata items.

### Step 1: getting the data

> The data is CC-BY but it was agreed that adding the SwissLipid identifiers to Wikidata (CCZero) is okay.

Reviewer: ```suggestion
The data is licensed under [CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/) but it was agreed (QUESTION@Egon: by whom? what was the process?) that adding the SwissLipid identifiers to Wikidata, (which is licensed under [CC0](https://creativecommons.org/publicdomain/zero/1.0/)) is fine.
```

Author: That was not in writing, but by the SIB team.

Author: Resolved this in a new commit: https://github.com/FAIRplus/the-fair-cookbook/pull/424/commits/d658a0e49451fdecb7ec1d55178283811c3b36d3

* Download `lipids.tsv` from the [Downloads](https://www.swisslipids.org/#/downloads) page
* Gunzip the file

### Step 2: extract Swiss Lipid ID <> InChIKey tuples

> For this step I use `csvtool` (`apt get install csvtool`):

Reviewer: ```suggestion
For this step, use `csvtool` (`apt get install csvtool`):
```

{footcite}`Waagmeester2021`.

### Step 4: creating QuickStatements

Now we have the mappings and the data model in Wikidata, we can create QuickStatements to allow us to enter the
data into Wikidata. This is not the only approach, and the process can be further automated using "Wikidata bots".
For this, see BioHackathon Europe [Project 32](https://github.com/elixir-europe/biohackathon-projects-2021/tree/main/projects/32):
Connecting ELIXIR-related open data on Wikidata via [WikiProject ELIXIR](https://www.wikidata.org/wiki/Wikidata:WikiProject_Elixir).

Based on existing Bacting scripts, a script is created to take the `swisslipids_ids.tsv` file as input and create
QuickStatements: [https://github.com/egonw/ons-wikidata/blob/master/ExtIdentifiers/swisslipids.groovy](https://github.com/egonw/ons-wikidata/blob/master/ExtIdentifiers/swisslipids.groovy) This script is using [Apache Groovy](http://www.groovy-lang.org/)
but Bacting can also be using in Python, see [https://github.com/cthoyt/pybacting](https://github.com/cthoyt/pybacting).

> This script uses a federated query against https://beta.sparql.swisslipids.org/sparql/ after a suggestion by Dr [Jerven Bolleman](https://orcid.org/0000-0002-7449-1266)
> who indicated that the [RDF4J](https://rdf4j.org/) backing this SPARQL endpoint will automatically batch the query against Wikidata, overcoming
> limitations of the Wikidata Query Service:

Reviewer: ````suggestion
```{note}
This script uses a federated query against https://beta.sparql.swisslipids.org/sparql/ after a suggestion by Dr [Jerven Bolleman](https://orcid.org/0000-0002-7449-1266)
who indicated that the [RDF4J](https://rdf4j.org/) backing this SPARQL endpoint will automatically batch the query against Wikidata, overcoming
limitations of the Wikidata Query Service:
```
````

Q76865004       P8691   "SLM:000163968" S248    Q41165322       S854    "https://www.swisslipids.org/#/downloads"       S813    +2021-11-06T00:00:00Z/11
Q76733356       P8691   "SLM:000164019" S248    Q41165322       S854    "https://www.swisslipids.org/#/downloads"       S813    +2021-11-06T00:00:00Z/11
Q76733312       P8691   "SLM:000164023" S248    Q41165322       S854    "https://www.swisslipids.org/#/downloads"       S813    +2021-11-06T00:00:00Z/11
Q76737210       P8691   "SLM:000164026" S248    Q41165322       S854    "https://www.swisslipids.org/#/downloads"       S813    +2021-11-06T00:00:00Z/11
Q76736881       P8691   "SLM:000164032" S248    Q41165322       S854    "https://www.swisslipids.org/#/downloads"       S813    +2021-11-06T00:00:00Z/11
Q76735022       P8691   "SLM:000164034" S248    Q41165322       S854    "https://www.swisslipids.org/#/downloads"       S813    +2021-11-06T00:00:00Z/11
```

This resulted in about 17.5 thousand mappings. These are based on exact InChIKey match. 

### Step 5: running the QuickStatements

The resulting mappings are being added via the [QuickStatements website](https://quickstatements.toolforge.org/).

## Conclusion

> It will be clear Bacting can be replaced without too much effort by other tools making the same steps. What is the
> main conclusion is that we can indeed automate adding external identifiers routinely to Wikidata using InChIKey
> matching.
> 
> What this protocol does not do, however, is create Wikidata items for lipids in the SwissLipids database that
> are not yet in Wikidata. This is the topic of a future recipe.
> 

Reviewer: ```suggestion
The main conclusion is that we can automate adding external identifiers routinely to Wikidata using InChIKey matching. 

It will be clear [Bacting tool](https://bio.tools/bacting) can be replaced without too much effort by other tools making the same steps.

What this protocol does not do, however, is create Wikidata items for lipids in the SwissLipids database that
are not yet in Wikidata. 

This is the topic of a future recipe.

```

## References

> 
> ```{footbibliography}
> ```
> 

Reviewer: ```suggestion
````{dropdown} **Reference**
```{footbibliography}
```
```

---

## Authors

> 
> ````{authors_fairplus}
> Egon: Writing - Original Draft
> ````
> 

Reviewer: ```suggestion

````{authors_fairplus}
Egon: Writing - Original Draft
Philippe:  Review & Editing
````

```

# Registering SwissLipids identifiers in Wikidata

 ````{panels_fairplus}

> :identifier_text: FCBxxx
> :identifier_link: 'https://w3id.org/faircookbook/FCBxxx'

Reviewer: ```suggestion
:identifier_text: FCB083
:identifier_link: 'https://w3id.org/faircookbook/FCB083'
```
